projects/sparse_rl/sparse_dqn.py

```diff
@@ -0,0 +1,372 @@
+# Copyright (C) 2019, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+from __future__ import print_function
+
+import argparse
+import collections
+import copy
+import csv
+import os
+import random
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+import tqdm
+
+from htmresearch.frameworks.pytorch.sparse_net import SparseNet
+from htmresearch.support.openai_utils import create_atari_environment
+
+SEED = 42
+SPARSE = True
+NUM_TIMESTEPS = int(1e6)
+ENVIRONMENT = "SeaquestDeterministic-v4"
+RESULTS_PATH = "results"
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+
+
+class Flatten(nn.Module):
+  """
+  Flatten input retaining batch dimension
+  """
+
+
+  def forward(self, x):
+    return x.view(x.size(0), -1)
+
+
+
+class DenseQ(nn.Module):
+  """
+  Neural Network described in https://arxiv.org/pdf/1312.5602.pdf
+  """
+
+
+  def __init__(self, actions):
+    super(DenseQ, self).__init__()
+
+    inputShape = (4, 84, 84)
+    self.model = nn.Sequential(
+      nn.Conv2d(in_channels=inputShape[0], out_channels=16, kernel_size=8,
+                stride=4),
+      nn.ReLU(),
+      nn.Conv2d(in_channels=16, out_channels=32, kernel_size=4, stride=2),
+      nn.ReLU(),
+      Flatten(),
+      nn.Linear(in_features=32 * 9 * 9, out_features=256),
+      nn.LayerNorm(256),
+      nn.ReLU(),
+      nn.Linear(in_features=256, out_features=actions),
+    )
+
+
+  def forward(self, x):
+    return self.model(x)
+
+
+
+class SparseQ(nn.Module):
+  """
+  Sparse version of DenseQ network
+  """
+
+
+  def __init__(self, actions):
+    super(SparseQ, self).__init__()
+
+    self.model = SparseNet(inputSize=(4, 84, 84),
+                           outChannels=[16, 32],
+                           kernelSize=[8, 4],
+                           stride=[4, 2],
+                           n=256,
+                           outputSize=actions,
+                           c_k=[160, 320], k=40, weightSparsity=0.3,
+                           useBatchNorm=True,
+                           useSoftmax=False)
+
+
+  def forward(self, x):
+    return self.model(x)
+
+
+
+class DQN(object):
```

Comment:
This looks pretty cool, and general. Should we create a frameworks/rl and move the DQN class and support stuff there into separate files? Could potentially move the open AI utils there too.

Reply:
Done

Reply:
Thanks!